src/analysis/processing/qgsoverlayutils.cpp

```diff
@@ -0,0 +1,118 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsoverlayutils.h"
+
+#include "qgsgeometryengine.h"
+#include "qgsprocessingalgorithm.h"
+
+///@cond PRIVATE
+
+void QgsOverlayUtils::difference( const QgsFeatureSource &sourceA, const QgsFeatureSource &sourceB, QgsFeatureSink &sink, QgsProcessingContext &context, QgsProcessingFeedback *feedback, int &count, int totalCount, QgsOverlayUtils::DifferenceOutput outputAttrs )
+{
+  QgsFeatureRequest requestB;
+  requestB.setSubsetOfAttributes( QgsAttributeList() );
+  if ( outputAttrs != OutputBA )
+    requestB.setDestinationCrs( sourceA.sourceCrs(), context.transformContext() );
+  QgsSpatialIndex indexB( sourceB.getFeatures( requestB ), feedback );
+
+  int fieldsCountA = sourceA.fields().count();
+  int fieldsCountB = sourceB.fields().count();
+  QgsAttributes attrs;
+  attrs.resize( outputAttrs == OutputA ? fieldsCountA : ( fieldsCountA + fieldsCountB ) );
+
+  if ( totalCount == 0 )
+    totalCount = 1;  // avoid division by zero
+
+  QgsFeature featA;
+  QgsFeatureRequest requestA;
+  if ( outputAttrs == OutputBA )
+    requestA.setDestinationCrs( sourceB.sourceCrs(), context.transformContext() );
+  QgsFeatureIterator fitA = sourceA.getFeatures( requestA );
+  while ( fitA.nextFeature( featA ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    if ( featA.hasGeometry() )
+    {
+      QgsGeometry geom( featA.geometry() );
+      QgsFeatureIds intersects = indexB.intersects( geom.boundingBox() ).toSet();
+
+      QgsFeatureRequest request;
+      request.setFilterFids( intersects );
+      request.setSubsetOfAttributes( QgsAttributeList() );
+      if ( outputAttrs != OutputBA )
+        request.setDestinationCrs( sourceA.sourceCrs(), context.transformContext() );
+
+      std::unique_ptr< QgsGeometryEngine > engine;
+      if ( !intersects.isEmpty() )
+      {
+        // use prepared geometries for faster intersection tests
+        engine.reset( QgsGeometry::createGeometryEngine( geom.constGet() ) );
+        engine->prepareGeometry();
+      }
+
+      QVector<QgsGeometry> geometriesB;
+      QgsFeature featB;
+      QgsFeatureIterator fitB = sourceB.getFeatures( request );
+      while ( fitB.nextFeature( featB ) )
+      {
+        if ( feedback->isCanceled() )
+          break;
+
+        if ( engine->intersects( featB.geometry().constGet() ) )
+          geometriesB << featB.geometry();
+      }
+
+      if ( !geometriesB.isEmpty() )
+      {
+        QgsGeometry geomB = QgsGeometry::unaryUnion( geometriesB );
+        geom = geom.difference( geomB );
```

Comment:
Do we need a wkb type check here? E.g. if input layer is polygons, throw away points/lines (is that even possible? Maybe not...). But definitely we need multipart handling at least - e.g. polygon layer which results in multipolygon results (split to multiple features? Abort with error?)

Reply:
> Do we need a wkb type check here?

Compared to intersection, we can't get a mixed bag of geometries from difference (AFAIK). We can only erase the whole geometry by subtracting "geomB" that completely covers "geom" - a case that we should actually handle properly and skip such feature (right now it is skipped with an error)

> But definitely we need multipart handling at least

This is already handled just like in the intersection algorithm - the sink is created as a multi-part geometry type. Test data produce both polygons and multi-polygons. Maybe we could add `geom.convertToMultiType()` to be safe with stricter data providers?

SAGA has an option to split multi-part geometries on output into single-part geometries. We could have such option at some point too.

Reply:
Ok, +1 to force multipart via convertToMultipart then